File: src/visiondrop/engine.py
"""The interaction engine: landmarks in, cursor + gesture state out.

This module is UI-free and platform-free on purpose. The live app, the debug
HUD, and the replay tests all drive the same ``Engine.process``.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from . import config as _config
from .cursor import Cursor
from .features import HandFeatures
from .filters import VelocityTracker
from .gestures import PinchEventKind, PinchFSM, PinchState, is_idle_pose
from .tracking import FrameObservation


def primary_screen_size() -> tuple[int, int]:
    """Screen size in pixels, falling back to 1920x1080 when unavailable."""
    try:
        import Quartz

        display = Quartz.CGMainDisplayID()
        return (
            int(Quartz.CGDisplayPixelsWide(display)),
            int(Quartz.CGDisplayPixelsHigh(display)),
        )
    except Exception:
        return (1920, 1080)


@dataclass(frozen=True)
class EngineState:
    timestamp_s: float
    hand_present: bool
    active: bool
    idle: bool
    features: HandFeatures | None
    pinch: PinchState | None
    cursor: tuple[int, int]
    velocity: float

    @property
    def events(self) -> tuple[PinchEventKind, ...]:
        return self.pinch.events if self.pinch else ()


class Engine:
    def __init__(
        self,
        config: _config.EngineConfig | None = None,
        screen_size: tuple[int, int] | None = None,
        frame_aspect: float = 16 / 9,
    ) -> None:
        self.config = config or _config.DEFAULT_CONFIG
        self.frame_aspect = frame_aspect
        self.cursor = Cursor(
            screen_size=screen_size or primary_screen_size(),
            config=self.config.cursor,
            filter_config=self.config.filters,
        )
        self.pinch = PinchFSM(self.config.pinch)
        self._velocity = VelocityTracker()

    def reset(self) -> None:
        self.pinch.reset()
        self.cursor.reset()
        self._velocity.reset()

    def process(self, observation: FrameObservation) -> EngineState:
        ts_s = observation.timestamp_s
        ts_ms = ts_s * 1000.0

        if not observation.hands:
            self.pinch.reset()
            return EngineState(
                timestamp_s=ts_s,
                hand_present=False,
                active=False,
                idle=True,
                features=None,
                pinch=None,
                cursor=self.cursor.position,
                velocity=0.0,
            )

        hand = max(observation.hands, key=lambda item: item.score)
        features = HandFeatures.from_landmarks(hand.landmarks, aspect=self.frame_aspect)
        speed = self._velocity(features.pinch_point, ts_s)
        idle = is_idle_pose(features, speed)

        ratio = math.inf if idle else features.pinch_ratio
        screen_pos = self.cursor.map_to_screen(features.pinch_point)
        pinch = self.pinch.update(ratio, screen_pos, ts_ms)

        active = features.pointing or pinch.closed
        if active:
            cursor_pos = self.cursor.update(features.pinch_point, ts_s, ts_ms)
        else:
            cursor_pos = self.cursor.position

        for event in pinch.events:
            if event is PinchEventKind.PRESS:
                self.cursor.freeze(ts_ms)
            elif event is PinchEventKind.DRAG_START:
                self.cursor.unfreeze()
            elif event in (PinchEventKind.CLICK, PinchEventKind.DOUBLE_CLICK):
                self.cursor.freeze(ts_ms)

        return EngineState(
            timestamp_s=ts_s,
            hand_present=True,
            active=active,
            idle=idle,
            features=features,
            pinch=pinch,
            cursor=cursor_pos,
            velocity=speed,
        )

File: Sources/VisionDropCore/Engine.swift
/// Everything the engine knows after one frame.
public struct EngineState: Sendable, Equatable {
    public let timestamp: Duration
    public let handPresent: Bool
    /// True when the hand is in a pose that drives the pointer.
    public let isActive: Bool
    /// True when the hand is resting; all input is suppressed (SRS ENG-11).
    public let isIdle: Bool
    public let features: HandFeatures?
    public let pinch: PinchState?
    public let pointer: ScreenPoint
    /// Continuous pinch closure, 0…1, for the overlay meter (SRS ENG-9).
    public let closure: Double
    /// Smoothed hand speed, normalized units per second.
    public let speed: Double

    public var events: [PinchEvent] { pinch?.events ?? [] }
}

/// The interaction engine: landmarks in, pointer and gesture state out.
///
/// UI-free and platform-free on purpose. The live app, the replay tool and the
/// tests all drive this same `process(_:)`, which is what makes a recorded
/// session a regression test (SRS MNT-1, ENG-7).
public struct Engine: Sendable {
    public var config: EngineConfiguration

    /// Frame width divided by frame height, used to make normalized distances
    /// isotropic.
    public var frameAspect: Double

    public private(set) var pointer: PointerMap
    private var pinch: PinchMachine
    private var velocity = VelocityTracker()
    private var wasActive = false

    public init(
        config: EngineConfiguration = .init(),
        layout: DisplayLayout = .single,
        frameAspect: Double = 16.0 / 9.0
    ) {
        self.config = config
        self.frameAspect = frameAspect
        self.pointer = PointerMap(layout: layout, config: config.pointer, filterConfig: config.filter)
        self.pinch = PinchMachine(config: config.pinch)
    }

    public mutating func reset() {
        pinch.reset()
        pointer.reset()
        velocity.reset()
        wasActive = false
    }

    public mutating func process(_ observation: FrameObservation) -> EngineState {
        let time = observation.timestamp

        guard let hand = observation.primaryHand,
            let features = HandFeatures(hand, aspect: frameAspect, config: config.features),
            features.isComplete
        else {
            // No usable hand. Every machine returns to neutral so that nothing
            // is latched across the gap and a drag cannot stick (SRS ENG-6).
            pinch.reset()
            velocity.reset()
            wasActive = false
            return EngineState(
                timestamp: time,
                handPresent: !observation.hands.isEmpty,
                isActive: false,
                isIdle: true,
                features: nil,
                pinch: nil,
                pointer: pointer.currentPosition,
                closure: 0,
                speed: 0
            )
        }
        let speed = velocity.update(features.pinchPoint, at: time)
        let idle = IdlePose.isIdle(features: features, speed: speed, config: config.idle)
        let openHand = features.isOpenPalm && features.pinchRatio > config.idle.openPinchRatio

        if idle || (openHand && pinch.isClosed) {
            pinch.reset()
            wasActive = false
            return EngineState(
                timestamp: time,
                handPresent: true,
                isActive: false,
                isIdle: true,
                features: features,
                pinch: nil,
                pointer: pointer.currentPosition,
                closure: features.closure(config.pinch),
                speed: speed
            )
        }
        let ratio = features.pinchRatio
        let mapped = pointer.mapToScreen(features.pinchPoint)
        let pinchState = pinch.update(ratio: ratio, position: mapped, at: time)

        let isActive = features.isPointing || pinchState.isClosed

        if isActive && !wasActive {
            // Re-anchor rather than teleport (SRS PTR-6).
            pointer.engage(at: features.pinchPoint)
        }
        wasActive = isActive

        if isActive {
            pointer.update(features.pinchPoint, at: time)
        }

        for event in pinchState.events {
            switch event {
            case .press, .click, .doubleClick:
                pointer.freeze(at: time)
            case .dragStart:
                pointer.unfreeze()
            case .dragEnd:
                break
            }
        }

        return EngineState(
            timestamp: time,
            handPresent: true,
            isActive: isActive,
            isIdle: idle,
            features: features,
            pinch: pinchState,
            pointer: pointer.currentPosition,
            closure: features.closure(config.pinch),
            speed: speed
        )
    }
}
